### PR TITLE
Speed up `ptr::replace`

### PR DESCRIPTION
**Pull request by lukas-code**
I noticed, that `ptr::replace` and `mem::replace` are implemented differently, even though they should probably be the same.

Looking at the [goldbolt output](https://godbolt.org/z/15b5GY883), it seems that `mem::replace` does only 2 `memcpy`s and `ptr::replace` does 2 `memcpy`s with an inlined copy in between.

I ran some benchmarks on my machine, in which `mem::replace` was more than 2x faster than `ptr::replace`.
<details>
<summary>Benchmark</summary>

### Code
```rust
#![feature(test)]
#![cfg(test)]

extern crate test;

use rand::Rng;
use std::{mem, ptr};
use test::Bencher;

#[bench]
fn mem_replace(b: &mut Bencher) {
    let mut rng = rand::thread_rng();

    let mut dst = [0u8; 256];
    rng.fill(&mut dst);

    let mut src = [0u8; 256];
    rng.fill(&mut src);

    b.iter(|| {
        let mut dst_copy = dst;
        let old = mem::replace(&mut dst_copy, src);
        (old, dst_copy)
    });
}

#[bench]
fn ptr_replace(b: &mut Bencher) {
    let mut rng = rand::thread_rng();

    let mut dst = [0u8; 256];
    rng.fill(&mut dst);

    let mut src = [0u8; 256];
    rng.fill(&mut src);

    b.iter(|| {
        let mut dst_copy = dst;
        let old = unsafe { ptr::replace(&mut dst_copy, src) };
        (old, dst_copy)
    });
}
```

### Output
```
test mem_replace ... bench:          11 ns/iter (+/- 0)
test ptr_replace ... bench:          25 ns/iter (+/- 0)
```
</details>

In this PR i have replaced the implementation of `ptr::replace` to call `mem::replace` instead of `mem::swap`.

### CODE REVIEW

**Comment by scottmcm** [comment]
cc @m-ou-se, who did https://github.com/rust-lang/rust/pull/83022 and thus probably has thoughts here.

also cc @RalfJung, who might have thoughts about the typed-ness of this like they did in https://github.com/rust-lang/rust/pull/97712

**Comment by RalfJung** [comment]
> also cc @RalfJung, who might have thoughts about the typed-ness of this like they did in https://github.com/rust-lang/rust/pull/97712

This operation takes and returns a `T`, so it has no choice but to do a typed copy.

It doesn't seem to actually help though?

File: library/core/src/ptr/mod.rs [hunk 1]
```diff
@@ -1012,16 +1012,15 @@ const unsafe fn swap_nonoverlapping_simple<T>(x: *mut T, y: *mut T, count: usize
 #[inline]
 #[stable(feature = "rust1", since = "1.0.0")]
 #[rustc_const_unstable(feature = "const_replace", issue = "83164")]
-pub const unsafe fn replace<T>(dst: *mut T, mut src: T) -> T {
+pub const unsafe fn replace<T>(dst: *mut T, src: T) -> T {
     // SAFETY: the caller must guarantee that `dst` is valid to be
     // cast to a mutable reference (valid for writes, aligned, initialized),
     // and cannot overlap `src` since `dst` must point to a distinct
     // allocated object.
```
Comment (RalfJung) [inline library/core/src/ptr/mod.rs:1019]:
```suggestion
    // and cannot overlap `src` since `src` is a freshly stack-allocated local variable.
```

**Comment by lukas-code** [comment]
It probably doesn't improve performance in rustc a lot (at all), because it barely seems to be used in the Rust repository. All uses of `ptr::replace` I could find are in the implementation of `mpsc` in the standard library.

https://github.com/rust-lang/rust/blob/7036449c774860a5b348dbbe01c20704c557382e/library/std/src/sync/mpsc/oneshot.rs#L182

https://github.com/rust-lang/rust/blob/7036449c774860a5b348dbbe01c20704c557382e/library/std/src/sync/mpsc/oneshot.rs#L216

https://github.com/rust-lang/rust/blob/7036449c774860a5b348dbbe01c20704c557382e/library/std/src/sync/mpsc/oneshot.rs#L295

https://github.com/rust-lang/rust/blob/7036449c774860a5b348dbbe01c20704c557382e/library/std/src/sync/mpsc/shared.rs#L260

https://github.com/rust-lang/rust/blob/7036449c774860a5b348dbbe01c20704c557382e/library/std/src/sync/mpsc/stream.rs#L164

Additionally I found one instance of `mem::replace` that should probably be a `ptr::replace`:
https://github.com/rust-lang/rust/blob/7036449c774860a5b348dbbe01c20704c557382e/library/std/src/thread/local.rs#L812-L831

File: library/std/src/thread/local.rs [hunk 2]
```diff
@@ -811,7 +811,7 @@ mod lazy {
 
             // SAFETY:
             //
-            // note that this can in theory just be `*ptr = Some(value)`, but due to
+            // Note that this can in theory just be `*ptr = Some(value)`, but due to
```
Comment (lukas-code) [inline library/std/src/thread/local.rs:814]: Due to what?

**Comment by JohnCSimon** [comment]
ping from triage:
@lukas-code  
looks like this PR is still in progress - so returning to author 

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

@rustbot author

**Comment by lukas-code** [comment]
Since this doesn't appear to increase performance at all, I'll just close this.

**Pull request closed without merging**